controllers/cloudstackfailuredomain_controller.go

```diff
@@ -130,21 +140,215 @@ func (r *CloudStackFailureDomainReconciliationRunner) ClearMachines() (ctrl.Resu
+				break
 			}
 		}
+		if etcdMachineFound {
+			break
+		}
 	}
-	if len(machines.Items) > 0 {
-		return r.RequeueWithMessage("FailureDomain still has machine(s) in it.")
+
+	return ctrl.Result{}, nil
+}
+
+func triggerMachineDeploymentRollout(machines []infrav1.CloudStackMachine, r *CloudStackFailureDomainReconciliationRunner) (ctrl.Result, error) {
+	workerMachineFound := false
+	for _, machine := range machines {
+		for _, ref := range machine.OwnerReferences {
+			if ref.Kind == "MachineSet" {
+				workerMachineFound = true
+
+				// get machine deployment by using cloudstack machine's label info
+				md := &clusterv1.MachineDeployment{}
+				mdName, ok := machine.Labels[clusterv1.MachineDeploymentLabelName]
+				if !ok {
+					return ctrl.Result{}, errors.Errorf("cloudstack machine %s misses label: %s", machine.Name, clusterv1.MachineDeploymentLabelName)
+				}
+
+				if err := r.K8sClient.Get(r.RequestCtx, client.ObjectKey{Namespace: machine.Namespace, Name: mdName}, md); err != nil {
+					return ctrl.Result{}, err
+				}
+
+				// add an annotation restartedAt in machine deployment if such one not already added
+				// this will trigger an immediate machine deployment rollout
+				_, ok = md.Spec.Template.Annotations["cluster.x-k8s.io/restartedAt"]
+				if !ok {
+					if md.Spec.Template.Annotations == nil {
+						md.Spec.Template.Annotations = map[string]string{}
+					}
+					timeNowStr := time.Now().Format(time.RFC3339)
+					md.Spec.Template.Annotations["cluster.x-k8s.io/restartedAt"] = timeNowStr
+					patcher, err := patch.NewHelper(md, r.K8sClient)
+					if err != nil {
+						return ctrl.Result{}, err
+					}
+					if err := patcher.Patch(r.RequestCtx, md); err != nil {
+						return ctrl.Result{}, err
+					}
+					r.Log.Info(fmt.Sprintf("failuredomain delete: machine deployment name %s add annotation restartedAt %s", mdName, timeNowStr))
+				} else {
+					r.Log.Info(fmt.Sprintf("failuredomain delete: machine deployment name %s already has restartedAt annotation, skip", mdName))
+				}
+				break
+			}
+		}
+		if workerMachineFound {
+			break
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func triggerControlPlaneRollout(machines []infrav1.CloudStackMachine, r *CloudStackFailureDomainReconciliationRunner) (ctrl.Result, error) {
+	var cpMachine infrav1.CloudStackMachine
+	var ownerRef metav1.OwnerReference
+	cpMachineFound := false
+	for _, machine := range machines {
+		for _, ref := range machine.OwnerReferences {
+			if ref.Kind == "KubeadmControlPlane" {
+				cpMachineFound = true
+				cpMachine = machine
+				ownerRef = ref
+			}
+			// etcdadmcluster rollout will trigger control plane rollout automatically.
+			// if etcd VM exists, trigger etcdadmcluster rollout is enough, no need to trigger control plane VM rollout.
+			if ref.Kind == "EtcdadmCluster" {
+				return ctrl.Result{}, nil
+			}
+		}
+	}
+	if !cpMachineFound {
+		return ctrl.Result{}, nil
```

Comment:
do we need to log anything here?

Reply:
No need to log anything since no CP VM in this failure domain is a normal situation. It is not an error or exceptional thing. 
I used to have something logged in this PR's draft version for easier search of logs.